### PR TITLE
Add starknet support for SAG

### PR DESCRIPTION
**Pull request by LeJamon**
Add a new SAG package tailored for starknet (starknet-sag-ts). 
This package is made to suits an on-chain verification in cairo and utilises Garaga MSM hint to do so.

### CODE REVIEW

**Review by Elli610 — COMMENTED** [review]
I guess the `coverage` folder should be added to the `.gitignore`

**Comment by Elli610** [comment]
Can you remove all the `.DS_Store` files ?

**Comment by Elli610** [comment]
I think we should rename the `starknet-sag-ts` to `cairo-sag-ts` to be more accurate. What do you think ?

**Comment by Elli610** [comment]
The tests for `./packages/starknet-sag-ts/src/ringSignature.ts` are missing

**Comment by Elli610** [comment]
we also need a read me to understand the purpose of the new package (if we do not merge it with `sag-ts`) and how we can run the code

**Comment by Elli610** [comment]
Looks like you forgot to add the commands for this new package in the makefile

**Comment by LeJamon** [comment]
> I think we should rename the `starknet-sag-ts` to `cairo-sag-ts` to be more accurate. What do you think ?

We should keep the name "starknet-sag-ts" rather than changing to "cairo-sag-ts". The implementation is specific to Starknet, not just Cairo in general. It includes Starknet-specific features like calldata generation. The current name accurately reflects its purpose and target platform.

**Comment by LeJamon** [comment]
> It looks like the `./packages/starknet-sag-ts/src/ringSignature.ts` file have a lot in common with the ` ./packages/sag-ts/ringSignature.ts`. Why didn't you added a value in the config as we did with the evm signatures ?

While they share common logic, combining them with a config approach (as done with EVM signatures) would result in overly complex code structures with numerous switch cases or nested if statements within functions. This would likely reduce code readability and maintainability.

**Comment by LeJamon** [comment]
> I guess the `coverage` folder should be added to the `.gitignore`

done

**Comment by LeJamon** [comment]
> we also need a readme to understand the purpose of the new package (if we do not merge it with `sag-ts`) and how we can run the code

done

**Comment by LeJamon** [comment]
> Looks like you forgot to add the commands for this new package in the makefile

done

**Comment by Elli610** [comment]
Can you remove the package.lock file ? We should only have a yarn.lock

**Comment by Elli610** [comment]
some DS_STORE remains (ex: packages/.DS_Store)

File: packages/starknet-sag-ts/src/signature/piSignature.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
+/*
+  This TypeScript library is the exclusive property of Cypher Lab (https://www.cypherlab.fr/) 
+  and is exclusively reserved for the use of gemWallet. Any form of commercial use, including but 
+  not limited to selling, licensing, or generating revenue from this code, is strictly prohibited.
+*/
```
Comment (Elli610) [inline packages/starknet-sag-ts/src/signature/piSignature.ts:5]:
```suggestion
```

File: packages/starknet-sag-ts/src/starknet-utils/cairo-utils.ts [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+ * resulting hash is compatible between Cairo (StarkNet) and TypeScript.
+ * @param {bigint[]} data - An array of bigint numbers representing the input to be hashed.
+ * @returns {bigint} - The resulting Keccak-256 hash represented as a bigint.
+ * @example
+ * const hash = cairoHash([123456789n, 987654321n]);
+ * console.log(hash); // Outputs a bigint hash
+ */
+export function cairoHash(data: bigint[]): bigint {
+  return BigInt("0x" + Buffer.from(poseidonHashMany(data).toString(16)));
+}
```
Comment (Elli610) [inline packages/starknet-sag-ts/src/starknet-utils/cairo-utils.ts:32]: description does not match the method content

File: packages/starknet-sag-ts/src/starknet-utils/garaga-utils.ts [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+        return [x, y];
+      });
+    case CurveName.SECP256K1:
+      return points.map((point) => point.toCoordinates());
+    default:
+      throw new Error(`Curve ${curveName} not supported`);
+  }
+}
+
+// eslint-disable-next-line
```
Comment (Elli610) [inline packages/starknet-sag-ts/src/starknet-utils/garaga-utils.ts:181]:
```suggestion
/* eslint-disable-next-line @typescript-eslint/no-explicit-any */
```

**Review by Elli610 — CHANGES_REQUESTED** [review]
ci need to be fixed

**Comment by LeJamon** [comment]
> Can you remove the package.lock file ? We should only have a yarn.lock

done

**Comment by LeJamon** [comment]
> some DS_STORE remains (ex: packages/.DS_Store)

done

**Comment by LeJamon** [comment]
> ci need to be fixed

done

**Comment by Elli610** [comment]
> > some DS_STORE remains (ex: packages/.DS_Store)
> 
> done

there are others files like this (ex: packages/starknet-sag-ts/src/signature/.DS_Store)

**Comment by LeJamon** [comment]
> > > some DS_STORE remains (ex: packages/.DS_Store)
> > 
> > 
> > done
> 
> there are others files like this (ex: packages/starknet-sag-ts/src/signature/.DS_Store)

I have deleted all of them

**Review by Elli610 — APPROVED** [review]
lgtm

**Pull request merged**